src/patch/apply.js

```diff
@@ -42,7 +40,61 @@ export function applyPatch(source, uniDiff, options = {}) {
             return false;
           }
         }
+        toPos++;
       }
+    }
+
+    return true;
+  }
+
+ // Search best fit offsets for each hunk based on the previous ones
+ for (let i = 0; i < hunks.length; i++) {
+    let hunk = hunks[i],
+        outOfLimits = 0,
+        localOffset = 0,
+        minLine = 0,
+        toPos = offset + hunk.oldStart - 1;
+
+    for (;;) {
+      if (toPos + localOffset + hunk.oldLines <= lines.length) {
+        if (hunkFits(hunk, toPos + localOffset)) {
+          hunk.offset = offset += localOffset;
+          break;
+        }
+      } else {
+        outOfLimits++;
+      }
+
+      // If trying to fit hunk outside both limits, return error
+      if (outOfLimits === 2) {
+        return false;
+      }
+
+      outOfLimits = 0;
+      localOffset++;
+
+      if (minLine <= toPos - localOffset) {
```

Comment:
You are right, that's the intended behaviour. The idea is that the fit should be near the original location, but we don't know if it's before or after it. Also, I'm taking in account the previous offset, since the line location will affect the next hunks too.